src/rest_framework_dso/embedding.py

```diff
@@ -7,89 +7,212 @@
 from __future__ import annotations
 
 from collections import defaultdict
+from dataclasses import dataclass
+from functools import cached_property
 from itertools import islice
-from typing import Callable, Dict, Iterable, Iterator, List, Optional, TypeVar, Union
+from typing import Callable, Dict, Iterable, Iterator, List, Optional, Type, TypeVar, Union
 
 from django.db import models
 from lru import LRU
 from rest_framework import serializers
-from rest_framework.exceptions import ParseError, PermissionDenied
+from rest_framework.exceptions import ParseError
 
-from dso_api.dynamic_api.permissions import fetch_scopes_for_model
-from rest_framework_dso.fields import AbstractEmbeddedField, EmbeddedField
+from rest_framework_dso.fields import AbstractEmbeddedField
 from rest_framework_dso.serializer_helpers import ReturnGenerator, peek_iterable
 
-EmbeddedFieldDict = Dict[str, AbstractEmbeddedField]
+DictOfDicts = Dict[str, Dict[str, dict]]
 T = TypeVar("T")
 M = TypeVar("M", bound=models.Model)
 DEFAULT_SQL_CHUNK_SIZE = 2000  # allow unit tests to alter this.
 
 
-def get_expanded_fields(  # noqa: C901
+def parse_expand_scope(
+    expand: Optional[str], expand_scope: Optional[str]
+) -> Union[bool, List[str]]:
+    """Parse the raw request parameters"""
+
+    # Initialize from request
+    if expand == "true":
+        # ?_expand=true should export all fields, unless `&_expandScope=..` is also given.
+        return expand_scope.split(",") if expand_scope else True
+    elif expand == "false":
+        return False
+    elif expand:
+        raise ParseError(
+            "Only _expand=true|false is allowed. Use _expandScope to expand specific fields."
+        ) from None
+
+    # Backwards compatibility, also allow `?_expandScope` without stating `?_expand=true`
+    # otherwise, parse as a list of fields to expand.
+    return expand_scope.split(",") if expand_scope else False
+
+
+def get_nested_fields_to_expand(
+    serializer_class: Type[serializers.Serializer],
+    expand_scope: Union[bool, List[str]],
+    allow_m2m=True,
+) -> DictOfDicts:
+    """Convert the string value of ?_expand/_expandScope to a browsable tree."""
+    if not expand_scope:
+        return {}
+    elif expand_scope is True:
+        return get_all_embedded_field_names(serializer_class, allow_m2m=allow_m2m)
+    else:
+        return group_dotted_names(expand_scope)
+
+
+@dataclass
+class EmbeddedFieldMatch:
```

Comment:
Nicely packaged into a dataclass! Makes it very readable indeed.

Reply:
:-)